text/0162-refactoring-support.md

```diff
@@ -0,0 +1,586 @@
+controlled history file on every commit. This event includes all creates,
+updates and deletes, plus all refactors, whether they were automatically
+detected or manually resolved.
+
+As with any event sourcing system, if we want to produce a snapshot of the
+stacks at a given point in time, all we need to do is replay the events in
+order, up to that point. We are now ready to state the key invariant of this
+system:
+
+> **Invariant**: For every revision `r`, the cloud assembly synthesized from
+> `r` is equal to the snapshot at `r`.
+
+In other words, the current state should be consistent with the history that led
+up to that state.
+
+One final piece to add to the system: every environment should also have its own
+history file, which should also maintain a similar invariant (through CFN hooks,
+for example). Having all this in place, we can execute the following algorithm
+on every deployment:
+
+    ---------------------------------
+      Key: 
+        H(E): environment history
+        H(A): application history
+        LCA: lowest common ancestor
+    ---------------------------------
+
+    if H(E) is a prefix of H(A):
+      Compute the diff between H(A) and H(E);
+      Extract the mapping from the diff;
+      Apply the mapping to the stacks in the environment;
+      Deploy;
+    else:
+      a = LCA of H(A) and H(E);
+      Compute the sub-chain of H(A) from a to the end;
+      Extract the mapping from the diff;
+      if the mapping is empty:
+        Deploy;
+      else:
+        Error: source state doesn't match the mapping.
+
+For example, suppose the histories at play are (`*` denotes the current state):
+
+    H(E) = e1 ◄── e2*
+    H(A) = e1 ◄── e2 ◄── e3 ◄── e4
+
+Then the diff between them is `e3 ◄── e4`. If these events contain any refactor,
+we just apply them, and then deploy the application. The resulting environment
+history is the merge of the two:
+
+    H(E) = e1 ◄── e2 ◄── e3 ◄── e4*
+
+Now suppose the histories involved are:
+
+    H(E) = e1 ◄── e2 ◄── e3*
+    H(A) = e1 ◄── e2 ◄── e4 ◄── e5
+
+In this case, `H(E)` is not a prefix of `H(A)`, but they have common ancestors.
+Their LCA is `e2`. Computing the sub-chain from there we get `e4 ◄── e5`. If
+there are no refactors to apply from this diff, we can go ahead and deploy the
+application. Again, the new state results from the merge of `H(E)` and
+`H(A)`:
+
+    H(E) = e1 ◄── e2 ◄── e3         
+                   ▲                  
+                   │                  
+                   └──── e4 ◄── e5*
+
+If there are refactors to be done, this is considered an error, because we can't
+guarantee that the refactor makes sense (let alone that this was the developer's
+intent).
+
+#### The future
+
+There is still some work to be done to prove this system works in practice. But
+assuming it does, we could use it to expand the scope to which the automatic
+refactoring applies. Consider the case in which you want to rename a certain
```

Comment:
ah so the proposed in-scope refactoring solution will treat this case (rename a resource + update properties of the resource) a replacement not refactoring? would we print out a warning message for this case?

Reply:
>ah so the proposed in-scope refactoring solution will treat this case (rename a resource + update properties of the resource) a replacement not refactoring?

Yes.

>would we print out a warning message for this case?

We could do that, but this case is not as clear-cut as it sounds. Unlike the in-scope case, there is no unambiguous way to differentiate between _refactor_ and _delete + creation_. We would need to define some distance function between two resources, and empirically come up with a threshold for that distance. If there are pairs whose distance is below the threshold, we would print a warning. But it's possible, and something I'm interested in exploring, anyway. This could be a phase 3.